### PR TITLE
CAD-3027 Merging p2p-master with master (Diffusion layer factorization)

### PR DESCRIPTION
**Pull request by bolt12**
This PR factors the `Diffusion` Layer in `P2P` and `NonP2P` and makes the necessary changes in both `ouroboros-network` and `ouroboros-consensus` repos, in order to enable a switch between starting a node as `p2p` or `non-p2p`

### CODE REVIEW

**Review by coot — COMMENTED** [review]
I see that one part of complexity that is pushed onto `ouroboros-consensus` comes from the difference between `DiffusionApplications` for non p2p and p2p versions, there are two parts that contribute to this:

1. different types in application fields
2. different auxiliaty members, e.g. `daErrorPolicies` vs `daRethrowPolicy`, and various additional fields in `p2p`: e.g. `daMiniProtocolParameters`, `daPeerMetrics`, `daBlockFetchMode`

To address 1. it could help if we made the non-p2p to use the same types as the p2p version
To address 2. we could either split the `DiffusionArguments` or provide a polymorphic member which would be instantiated to different types for p2p and non p2p. Whatever would make the use of the api simpler.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:228]: I tried to keep it as `Void` but it is not possible... so, in order to make the least decisions possible I just kept everything as `Either (nonp2p type) (p2p type)`

**Comment by bolt12** [comment]
> I see that one part of complexity that is pushed onto `ouroboros-consensus` comes from the difference between `DiffusionApplications` for non p2p and p2p versions, there are two parts that contribute to this:
> 
>     1. different types in application fields
> 
>     2. different auxiliaty members, e.g. `daErrorPolicies` vs `daRethrowPolicy`, and various additional fields in `p2p`: e.g. `daMiniProtocolParameters`, `daPeerMetrics`, `daBlockFetchMode`
> 
> 
> To address 1. it could help if we made the non-p2p to use the same types as the p2p version
> To address 2. we could either split the `DiffusionArguments` or provide a polymorphic member which would be instantiated to different types for p2p and non p2p. Whatever would make the use of the api simpler.

I tried to make the least decisions possible and to uniform everything between p2p and non-p2p. However, I figured the difference in the types would be too big and the needed modifications would be too complex. I like the way I have things because I believe it will make it easier to revert back once we decide to only keep the P2P part!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:236]: I'd suggest to use `()` instead, we'll switch to `Void` when it will be possible.

**Comment by coot** [comment]
> I tried to make the least decisions possible and to uniform everything between p2p and non-p2p. However, I figured the difference in the types would be too big and the needed modifications would be too complex. I like the way I have things because I believe it will make it easier to revert back once we decide to only keep the P2P part!

The most difficult part of this would be to modify the `non-p2p` part.  I am not 100% if this would work, but I'd like to experiment with it.  If it does work, we should be able to have PR to `master` which prepares that branch for merging `p2p-master` (it would include some other commits from `p2p-master` branch like  364a18ce6 and cb54ad7c2).  The modifications of `p2p`, i.e. the split of `DiffusionApplications` could stay with us for longer.

**Comment by bolt12** [comment]
Hmm, I guess I can try open a PR against master that tries to achieve that, it might end up being simpler than I am imagining! 😄

**Comment by coot** [comment]
> Hmm, I guess I can try open a PR against master that tries to achieve that, it might end up being simpler than I am imagining! smile

To make sure we don't have a complex conflict, make  a PR against `p2p-master` first, later we'll try to rewrite the history of `p2p-master` so this lands at the bottom of it (this is where a conflict might surface).

**Comment by bolt12** [comment]
@coot `P2P` and `Non-P2P` `runDataDiffusion` functions are very different, I was able to refactor `DiffusionTracers` and make everything run on `IO ()`, but `DiffusionApplications` and `DiffusionArguments` are very difficult to refactor, there are a lot of implementation details in each one, so changing the `non-p2p` part would have to introduce quite some of the `p2p` features. Do you have any suggestion or guidance that you can provide? I still think that making a clear distinction between the `p2p` and `non-p2p` for now is the simplest and cleanest approach, whenever we decide to make the full switch to `p2p`, we just need to remove all the `NonP2P.x` things

**Comment by bolt12** [comment]
**NOTE**: Before merging make sure that we are using the most recent versions of `P2P` ( `p2p-master` branch ) and `NonP2P` ( `master` branch ), and if not update the code.

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:335]: Let's align the types (`::`).

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:466]: This belongs to the previous section, but it's a mistake in the `p2p-master`, so let's leave it for now.

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion.hs:406]: We could eliminate it with a type level tag
```
data P2PMode = P2P | NonP2P
```
And then index all types with either `P2P` or `NonP2P`.  But maybe it is not worth it, I'll leave it up to. 

Hydra guys are also using data diffusion interface as well, we might need to help them to make a transition.

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:18]: good idea!

**Review by coot — CHANGES_REQUESTED** [review]
Very nice, some cosmetic remarks.

File: ouroboros-network/src/Ouroboros/Network/Diffusion/Common.hs
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/Common.hs:121]: sounds good to me! :D

File: ouroboros-network/src/Ouroboros/Network/Diffusion/NonP2P.hs
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/NonP2P.hs:70]: Hm, between `DiffusionTracersExt` and `DiffusionTracersExtra`, I prefer the latter, if it is okay having long names

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:335]: Got it!

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:406]: I don't think its worth it, the smart constructors will take care of everything and if someone uses the wrong smart constructors, I think the compiler can not type check

**Comment by bolt12** [comment]
Thanks @coot , will address your reviews!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:447]: I think consensus is using 2 space for indentation, rather than one.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:464]: The same here.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:99]: Please move these up with the other `Ouroboros.Network.*` imports.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:244]: Please add `enableP2P` as a field in `RunNodeArgs` instead of adding it as a new separate parameter to `run` and `runWith`.

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:244]: As a field, you can give it a polarized name, so it can just be a `Bool` if you wish. eg `rnEnableP2P :: Bool`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:414]: Would you add a comment here, explain what `mempty` is?

Or better yet, create a `where` binding

```
localRethrowPolicy :: RethrowPolicy
localRethrowPolicy = mempty
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:326]: Please undo these white-space changes.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:341]: Please undo these white-space changes.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:374]: Our style (link for reference https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus/docs/StyleGuide.md ) is

```
     Foo
  -> Bar
  -> Baz
```

so please re-introduce the leading space on the first argument.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:384]: Our style (link for reference https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus/docs/StyleGuide.md ) is

```
     Foo
  -> Bar
  -> Baz
```

so please re-introduce the leading space on the first argument.

**Review by nfrisby — CHANGES_REQUESTED** [review]
Just a few minor alternations requested to the Consensus files. Thanks!

File: ouroboros-consensus/ouroboros-consensus.cabal
Comment (bolt12) [reply ouroboros-consensus/ouroboros-consensus.cabal:241]: Will do! Thanks!

Comment (bolt12) [reply ouroboros-consensus/ouroboros-consensus.cabal:241]: Got it! Thanks :)

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:244]: I think I prefer having the `NetworkP2PMode` type, if that is okay

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:116]: Please put these back where they were. I know the reasoning is not obvious; please bear with me.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:211]: Please undo this line-break.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:116]: Builkite was failing because of this, I believe, that's why I did this

Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:116]: Should I ignore buildkite?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:211]: Got it!

Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:211]: hmm this seems to not be related to the newest state of the file

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:116]: Hmm. Please do respect BuildKite, yeah.

I'm guessing you removed the blank line (old line 105). I could see how that would invite a BuildKite error.

My request includes re-introducing that blank line too. To elaborate a bit the `Ouroboros.Consensus.Storage.*` imports have their own group, and it is separated by blank lines from the above group `Ouroboros.Consensus.*` imports.

Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:116]: Ah I see! Thanks for clarifying! Will do!

**Review by nfrisby — COMMENTED** [review]
Thanks Armando! There are a couple conversations still open from my first pass (about indenting the first argument type) and I made a couple more. But you already moved the new argument into the record type, and that was the only really important one in my opinion.

So I'll change my Review to "LGTM"! Thanks for coping with our style restrictions etc. Generally: if you make the least possible formatting changes, it'll be easier to get them right :)

The only reason I'm not Approving, is to ensure that others have the opportunity to review. 👍  Thanks again, and thank you for your patience re my review.

**Comment by bolt12** [comment]
Thank you @nfrisby ! Yes, thank you for pointing out the styling issues, I am more than happy to please the code owners, since I am manipulating their code base, so thank you as well for insisting on them!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:373]: Please change `(BlockNodeToNodeVersion blk` to `(   BlockNodeToNodeVersion blk`

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:383]: Please change `(BlockNodeToNodeVersion blk` to `(   BlockNodeToNodeVersion blk`

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:1245]:
```suggestion
      Right (sn, sd, _) -> Snocket.close sn sd
    )
````

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:219]:
```suggestion
data NetworkMode = P2PMode
                 | FederatedMode
  derivng (Eq, Ord)
```
Maybe let's use on slack for other ideas.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:336]:
```suggestion
    mkNodeToNodeApps
      :: NodeKernelArgs m (ConnectionId addrNTN) (ConnectionId addrNTC) blk

```

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:358]: The same here, let's not change the consensus style.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:371]: and here.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:219]: Maybe `Federated` and `P2P` are also nicer suffixes than `NonP2P` and `P2P` (but it would be a pervasive change).

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:358]: Yes, thank you!

**Review by nfrisby — COMMENTED** [review]
I marked two last style-y comments; quite minor.

Thanks much for restyling. I leave it to a Network Team member to proper Approve.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:382]: Got it! :)

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**
